celltype_annotation/analysis/SingleR-combining-refs.Rmd

```diff
@@ -0,0 +1,218 @@
+```
+
+
+```{r}
+# build path to annotated sce file 
+annotated_sce_file <- glue::glue("{params$library_id}_annotated.rds")
+local_annotated_sce_path <- file.path(params$local_data_dir, 
+                                      params$sample_id, 
+                                      annotated_sce_file)
+
+# if missing the annotated sce files, grab them from S3
+if(!all(file.exists(local_annotated_sce_path))){
+  # sync annotated SCE files 
+  aws_includes <- glue::glue('--include "*/{annotated_sce_file}"')
+  
+  sync_call <- glue::glue('aws s3 sync {params$s3_data_dir} {params$local_data_dir} --exclude "*" {aws_includes}')
+  system(sync_call, ignore.stdout = TRUE) 
+}
+
+# read in all annotated sce 
+annotated_sce <- readr::read_rds(local_annotated_sce_path)
+```
+
+## SingleR with combined reference
+
+Before providing the list of references and labels, first subset them so that they contain the same set of genes.
+
+```{r}
+# identify shared genes among all references
+shared_ref_genes <- celldex_refs |>
+    purrr::map(rownames) |>
+    purrr::reduce(intersect)
+
+# create list of refs with shared genes only 
+celldex_refs_shared <- celldex_refs |>
+  purrr::map(\(ref)
+             ref <- ref[shared_ref_genes ,])
+```
+
+
+```{r}
+# create list of labels 
+ont_labels <- celldex_refs_shared |> 
+  purrr::map(\(ref) ref$label.ont)
+```
+
+
+```{r}
+# perform SingleR using combined refs and labels
+combined_result <- SingleR::SingleR(test = annotated_sce, 
+                                    ref = celldex_refs_shared,
+                                    labels = ont_labels)
+```
+
+## Evaluate combined reference annotations
+
+Before evaluating some of the output, I wanted to be able to interpret the ontology ID's so the first thing I did was map the `label.ont` back to the main labels from each reference then combine that with the SingleR results.
+
+```{r}
+# for each reference, provide a map between ont labels and other labels in the dataset
+ref_labels <- celldex_refs_shared |>
+  purrr::map(\(ref) data.frame(label_ont = ref$label.ont,
+                               label_main = ref$label.main,
+                               label_fine = ref$label.fine)) |>
+  dplyr::bind_rows(.id = "label_reference")
+
+# add main and fine labels to combined results 
+result_df <- combined_result |>
+  as.data.frame() |>
+  # first recode the reference assigned 
+  dplyr::mutate(reference = dplyr::case_when(reference == 1 ~ "BlueprintEncodeData", 
+                                             reference == 2 ~ "HumanPrimaryCellAtlasData",
+                                             reference == 3 ~ "DatabaseImmuneCellExpressionData",
+                                             reference == 4 ~ "MonacoImmuneData")) |>
+  # then combine with labels to get main and fine labels associated with ont labels for each reference
+  dplyr::left_join(ref_labels, by = c("pruned.labels" = "label_ont", "reference" = "label_reference")) 
```

Comment:
Doing it this way I think kind of defeats the purpose of using `label.ont`; we really only want one label per "cell type" and then use the same labels across all refs for visualization. The best way to do that is to look up in Cell Ontology what the preferred name is for each ontology

We should be able to that do that pretty easily (adapting a bit from [chapter 6])(http://bioconductor.org/books/release/SingleRBook/exploiting-the-cell-ontology.html#exploiting-the-cell-ontology)

```suggestion
# probably do this higher up and separately
cl_ont <- ontoProc::getOnto('cellOnto')

# add on cell ontology names
result_df |> 
  mutate(label_name = cl_ont$name[pruned.labels])
```